File: voice_profiles/schemas.py
"""Schemas, enums and typed errors for the voice-profile backend core.

This module is the single source of truth for the JSON shape written to
``voice_profiles_data/{profile_id}/profile.json``. It deliberately mirrors
the conventions of :mod:`voice_clone.schemas` (pydantic v2 models, string
enums, hard constants) but does not import anything from the voice-clone
vertical slice: the voice-profile core must stay isolated.
"""

from __future__ import annotations

from enum import Enum
from typing import Any, Optional

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Schema versioning
# ---------------------------------------------------------------------------

SCHEMA_VERSION = 1
SUPPORTED_SCHEMA_VERSIONS: frozenset[int] = frozenset({1})

SUPPORTED_LOCALES: frozenset[str] = frozenset({"de-DE"})
DEFAULT_LOCALE = "de-DE"

MAX_PROFILE_NAME_LEN = 80
EXPECTED_PACK_PROMPT_COUNT = 88


# ---------------------------------------------------------------------------
# Enums
# ---------------------------------------------------------------------------

class ReferenceStatus(str, Enum):
    """Status of a reference attached to a profile."""

    ACCEPTED = "ACCEPTED"
    REVIEW = "REVIEW"
    REJECTED = "REJECTED"


class QualityClass(str, Enum):
    """Technical quality class reported by the voice-clone analyzer."""

    EXCELLENT = "EXCELLENT"
    GOOD = "GOOD"
    REVIEW = "REVIEW"
    REJECT = "REJECT"


# Mapping from a technical quality class to the canonical reference status.
QUALITY_CLASS_TO_STATUS: dict[QualityClass, ReferenceStatus] = {
    QualityClass.EXCELLENT: ReferenceStatus.ACCEPTED,
    QualityClass.GOOD: ReferenceStatus.ACCEPTED,
    QualityClass.REVIEW: ReferenceStatus.REVIEW,
    QualityClass.REJECT: ReferenceStatus.REJECTED,
}


# ---------------------------------------------------------------------------
# Typed errors
# ---------------------------------------------------------------------------

class VoiceProfileError(Exception):
    """Base class for all voice-profile core errors."""


class VoiceProfileValidationError(VoiceProfileError):
    """Hard validation failure (bad name, bad filename, bad quality, ...)."""


class VoiceProfileNotFoundError(VoiceProfileError):
    """A profile with the given id does not exist."""


class VoiceProfileConflictError(VoiceProfileError):
    """A profile-level conflict (e.g. trying to accept a REJECTED reference)."""


class VoiceScriptNotFoundError(VoiceProfileError):
    """A script id was not found in the loaded script library."""


class VoiceProfileStorageError(VoiceProfileError):
    """A persistence-layer failure (corrupt JSON, unknown schema, IO)."""


# ---------------------------------------------------------------------------
# Script library models
# ---------------------------------------------------------------------------

class RecommendedDuration(BaseModel):
    min: float
    max: float


class ScriptPrompt(BaseModel):
    """A single recording prompt loaded from the script library."""

    id: str
    order: int
    category: str
    style: str
    text: str
    recommended_duration_seconds: RecommendedDuration
    tags: list[str] = Field(default_factory=list)
    recording_notes: Optional[str] = None


class ScriptPack(BaseModel):
    """Top-level shape of a pack/holdout JSON file."""

    schema_version: int
    id: str
    locale: str
    name: str
    version: str
    expected_prompt_count: Optional[int] = None
    prompts: list[ScriptPrompt]


# ---------------------------------------------------------------------------
# Voice-profile models
# ---------------------------------------------------------------------------

class Progress(BaseModel):
    """Derived progress for a profile. Never persisted as a stale counter."""

    total: int
    missing: int
    recorded: int
    accepted: int
    review: int
    rejected: int
    percentage: float
    clone_ready: bool
    pack_complete: bool


class Reference(BaseModel):
    """A reference attached to a profile, keyed by stable script id."""

    script_id: str
    script_text: str
    category: str
    style: str
    recording_filename: str
    recording_sha256: str
    quality: dict[str, Any] = Field(default_factory=dict)
    quality_class: QualityClass
    status: ReferenceStatus
    review_accepted: bool = False
    attached_at: str
    updated_at: str


class Profile(BaseModel):
    """A persisted voice profile."""

    schema_version: int = SCHEMA_VERSION
    id: str
    name: str
    locale: str
    created_at: str
    updated_at: str
    archived: bool = False
    references: dict[str, Reference] = Field(default_factory=dict)

File: frontend/src/features/voiceProfiles/schemas.ts
import { z } from "zod";

/**
 * Voice Profile feature Zod schemas.
 *
 * Every API response is validated at runtime so the UI never silently
 * consumes unknown JSON shapes. The schemas mirror the actual backend
 * responses from the FastAPI voice-profile endpoints (see
 * voice_profiles_api.py / voice_profiles service). Unknown *additional*
 * fields are tolerated (Zod strips them by default), but missing required
 * fields and unknown enum values surface as controlled ApiErrors instead
 * of React crashes.
 *
 * Reference statuses follow the same permissive convention used for
 * generation statuses in `types/schemas.ts`: the schema accepts any string
 * so a future backend status renders as a neutral badge instead of
 * throwing. KNOWN_REFERENCE_STATUSES lists the values the UI understands.
 */

export const KNOWN_REFERENCE_STATUSES = [
  "ACCEPTED",
  "REVIEW",
  "REJECTED",
] as const;

export type KnownReferenceStatus = (typeof KNOWN_REFERENCE_STATUSES)[number];

/** Permissive status: known values get coloured badges, unknown ones neutral. */
export const referenceStatusSchema = z.string();

/** Quality class reported by the real voice_clone.quality analyzer. */
export const qualityClassSchema = z.string();

const recommendedDurationSchema = z.object({
  min: z.number(),
  max: z.number(),
});

/** A single recording prompt from the script pack / holdout. */
export const voiceScriptSchema = z.object({
  id: z.string(),
  order: z.number(),
  category: z.string(),
  style: z.string(),
  text: z.string(),
  recommended_duration_seconds: recommendedDurationSchema,
  tags: z.array(z.string()),
  recording_notes: z.string().nullable().optional(),
});

/** Pack metadata returned alongside the prompt list. */
export const voiceScriptPackMetaSchema = z.object({
  pack_id: z.string(),
  locale: z.string(),
  title: z.string().optional(),
  prompt_count: z.number(),
});

/** Response shape of GET /scripts and GET /holdout-scripts. */
export const voiceScriptPackSchema = z.object({
  pack: voiceScriptPackMetaSchema,
  prompts: z.array(voiceScriptSchema),
});

/** A reference attached to a profile. The backend stores references as a
 * dict keyed by script_id; the API returns that dict directly. */
export const voiceProfileReferenceSchema = z.object({
  script_id: z.string(),
  script_text: z.string(),
  category: z.string(),
  style: z.string(),
  recording_filename: z.string(),
  recording_sha256: z.string(),
  quality: z.record(z.string(), z.unknown()),
  quality_class: qualityClassSchema,
  status: referenceStatusSchema,
  review_accepted: z.boolean(),
  attached_at: z.string(),
  updated_at: z.string(),
});

/** Derived progress for a profile. Computed server-side from the script
 * library + stored references; never persisted as a stale counter. */
export const voiceProfileProgressSchema = z.object({
  total: z.number(),
  missing: z.number(),
  recorded: z.number(),
  accepted: z.number(),
  review: z.number(),
  rejected: z.number(),
  percentage: z.number(),
  clone_ready: z.boolean(),
  pack_complete: z.boolean(),
});

/** A persisted voice profile. References are a dict keyed by script_id. */
export const voiceProfileSchema = z.object({
  schema_version: z.number().optional(),
  id: z.string(),
  name: z.string(),
  locale: z.string(),
  created_at: z.string(),
  updated_at: z.string().optional(),
  references: z.record(z.string(), voiceProfileReferenceSchema),
  progress: voiceProfileProgressSchema,
});

export const voiceProfileListResponseSchema = z.object({
  profiles: z.array(voiceProfileSchema),
});

export const voiceProfileDeleteResponseSchema = z.object({
  id: z.string(),
  deleted: z.boolean(),
});

/** Best-effort error response schema. FastAPI errors return `{detail: ...}`. */
export const apiErrorSchema = z.object({
  detail: z.union([z.string(), z.array(z.any()), z.record(z.string(), z.any())]),
});
